examples/matrix-factorisation/SparseRatingMatrix.hs

```diff
@@ -0,0 +1,92 @@
+{-# LANGUAGE BlockArguments #-}
+{-# LANGUAGE FunctionalDependencies #-}
+{-# LANGUAGE RecordWildCards #-}
+
+module Main where
+
+import Control.Monad (foldM, when)
+import MF
+import Torch.Autograd (grad, makeIndependent, toDependent)
+import Torch.DType (DType (Float))
+import Torch.Functional (matmul)
+import Torch.Internal.Managed.Type.Context (manual_seed_L)
+import Torch.NN
+  ( Randomizable,
+    sample,
+    sgd,
+  )
+import Torch.Tensor (Tensor, asTensor, toDouble, toInt)
+import Torch.TensorFactories (randint', randn')
+
+type Item = Int
+
+type User = Int
+
+type RatingValue = Double
+
+data Rating = Rating Item User RatingValue deriving (Show)
+
+data Ratings = Ratings [Rating]
+
+data RatingsSpec
+  = RatingsSpec {usersNumber :: Int, itemsNumber :: Int, ratingNumber :: Int}
+
+instance Randomizable RatingsSpec Ratings where
+  sample (RatingsSpec n m l) = do
+    ratings <- randn' [1, l]
+    users <- randint' 1 n [1, l]
+    items <- randint' 1 m [1, l]
+    let list_ratings = toDouble <$> (getEntryByIndex ratings 0) <$> [0 .. (l -1)]
+        list_users = toInt <$> (getEntryByIndex users 0) <$> [0 .. (l -1)]
+        list_items = toInt <$> (getEntryByIndex items 0) <$> [0 .. (l -1)]
+    pure $ Ratings [Rating u i r | (u, i, r) <- zip3 list_users list_items list_ratings]
```

Comment:
while this is fine, IMO it would be more efficient if you defined a batched type, e.g.
```haskell
data RatingBatch = RatingBatch Items Users RatingValues deriving (Show)
newtype Items = Items Tensor deriving (Show)
newtype Users = Users Tensor deriving (Show)
newtype Ratings = Ratings Tensor deriving (Show)
```
the you wouldn't need to zip here but just assemble the data structure.

Reply:
Thanks @tscholak for your comment.
I just wanted to recall that I assumed that the batches are subsamples of our main data. I guess one of the sources of the slowness of the previous code is the ratings list size, and that is why we consider subsamples (batches) in each round of the gradient descent.
If we follow the above data types, for subsampling of `RatingBatch` we need to do it three times against `Items`, `Users` and `RatingValues` whilst we do it once if we zip them.
Please correct me if I'm wrong.